desktop/src/backends/navigator.rs

```diff
@@ -621,3 +621,231 @@ impl<F: FutureSpawner> NavigatorBackend for ExternalNavigatorBackend<F> {
+            let (socket, _) = listener.accept().or(async_timeout!()).await.unwrap();
+            socket
+        });
+        (accept_task, addr)
+    }
+
+    fn connect_test_socket(
+        addr: SocketAddr,
+        timeout: Duration,
+        socket_allow: bool,
+    ) -> (Sender<Vec<u8>>, Receiver<SocketAction>) {
+        let mut backend = new_test_backend(socket_allow);
+
+        let (write, receiver) = async_channel::unbounded();
+        let (sender, read) = async_channel::unbounded();
+
+        backend.connect_socket(
+            addr.ip().to_string(),
+            addr.port(),
+            timeout,
+            dummy_handle!(),
+            receiver,
+            sender,
+        );
+
+        (write, read)
+    }
+
+    async fn write_server(server_socket: &mut TcpStream, data: &str) {
+        server_socket
+            .write(data.as_bytes())
+            .or(async_timeout!())
+            .await
+            .expect("server write");
+    }
+
+    async fn read_server(server_socket: &mut TcpStream) -> String {
+        let mut buffer = [0; 4096];
+
+        let read = match server_socket.read(&mut buffer).await {
+            Err(e) if e.kind() == ErrorKind::TimedOut => {
+                panic!("TimedOut");
+            }
+            Err(_) | Ok(0) => {
+                panic!("closed");
+            }
+            Ok(read) => read,
+        };
+
+        let buffer = buffer.into_iter().take(read).collect::<Vec<_>>();
+        String::from_utf8(buffer).unwrap()
+    }
+
+    async fn write_client(client_write: &Sender<Vec<u8>>, data: &str) {
+        client_write
+            .send(data.as_bytes().to_vec())
+            .or(async_timeout!())
+            .await
+            .expect("client write");
+    }
+
+    #[macro_rules_attribute::apply(async_test)]
+    async fn test_socket_timeout() {
+        let (_accept_task, addr) = start_test_server().await;
+        let (_client_write, client_read) = connect_test_socket(addr, TIMEOUT_ZERO, true);
+        assert_next_socket_actions!(
+            client_read;
+            Connect(dummy_handle!(), ConnectionState::TimedOut),
+        );
+    }
+
+    #[macro_rules_attribute::apply(async_test)]
+    async fn test_socket_connect() {
+        let (accept_task, addr) = start_test_server().await;
+        let (_client_write, client_read) = connect_test_socket(addr, TIMEOUT, true);
+        let _server_socket = accept_task.await.unwrap();
+        assert_next_socket_actions!(
+            client_read;
+            Connect(dummy_handle!(), ConnectionState::Connected),
+        );
+    }
+
+    #[macro_rules_attribute::apply(async_test)]
+    async fn test_socket_deny() {
+        let addr = SocketAddr::from_str("127.0.0.42:65000").expect("dummy address");
+        let (_client_write, client_read) = connect_test_socket(addr, TIMEOUT, false);
+        assert_next_socket_actions!(
+            client_read;
+            Connect(dummy_handle!(), ConnectionState::Failed),
+        );
```

Comment:
Since this fails the same way if a server is not listening on that address, how about starting a server and making sure socket allow does what is expected.

Reply:
That's a good catch, should be good now :)